cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh

```diff
@@ -0,0 +1,149 @@
+#include <thrust/execution_policy.h>
+#include <thrust/fill.h>
+#include <thrust/memory.h>
+
+#include <iterator>
+#include <limits>
+#include <memory>
+#include <type_traits>
+
+namespace cuspatial {
+namespace detail {
+
+/**
+ * @internal
+ * @brief The kernel to compute point to linestring distance
+ *
+ * Each thread computes the distance between a line segment in the linestring and the
+ * corresponding point in the pair. The shortest distance is computed in the output
+ * array via an atomic operation.
+ *
+ * @tparam Cart2dItA Iterator to 2d cartesian coordinates. Must meet requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
+ * @tparam Cart2dItB Iterator to 2d cartesian coordinates. Must meet requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
+ * @tparam OffsetIterator Iterator to linestring offsets. Must meet requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
+ * @tparam OutputIterator Iterator to output distances. Must meet requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible and mutable.
+ *
+ * @param[in] points_first Iterator to the begin of the range of the points
+ * @param[in] points_last  Iterator to the end of the range of the points
+ * @param[in] linestring_offsets_begin Iterator to the begin of the range of the linestring offsets
+ * @param[in] linestring_offsets_end Iterator to the end of the range of the linestring offsets
+ * @param[in] linestring_points_begin Iterator to the begin of the range of the linestring points
+ * @param[in] linestring_points_end Iterator to the end of the range of the linestring points
+ * @param[out] distances Iterator to the output range of shortest distances between pairs.
+ *
+ * [LinkLRAI]: https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator
+ * "LegacyRandomAccessIterator"
+ */
+template <typename Cart2dItA, typename Cart2dItB, typename OffsetIterator, typename OutputIterator>
+void __global__ pairwise_point_linestring_distance(Cart2dItA points_first,
+                                                   OffsetIterator linestring_offsets_first,
+                                                   OffsetIterator linestring_offsets_last,
+                                                   Cart2dItB linestring_points_first,
+                                                   Cart2dItB linestring_points_last,
+                                                   OutputIterator distances)
+{
+  using T = iterator_vec_base_type<Cart2dItA>;
+
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x;
+       idx < std::distance(linestring_points_first, linestring_points_last);
+       idx += gridDim.x * blockDim.x) {
+    auto offsets_iter =
+      thrust::upper_bound(thrust::seq, linestring_offsets_first, linestring_offsets_last, idx);
+    // Pointer to the last point in the linestring.
+    if (*offsets_iter - 1 == idx) { return; }
```

Comment:
Each iteration takes care of one line segment. The last point of the linestring does not address any existing line segment, so should be skipped. Note that this line is updated since your review. See: https://github.com/rapidsai/cuspatial/pull/573#issuecomment-1218361393